validator/src/main/java/com/google/daq/mqtt/util/MessageUpgrader.java

```diff
@@ -48,30 +48,53 @@ public MessageUpgrader(String schemaName, JsonNode message) {
 
   /**
    * Update message to the latest standard.
+   *
+   * @param forceUpgrade true to force a complete upgrade pass irrespective of original version
+   * @return true if the message has been altered
    */
-  public void upgrade() {
+  public boolean upgrade(boolean forceUpgrade) {
     if (major != 1) {
       throw new IllegalArgumentException("Starting major version " + major);
     }
+
+    final JsonNode original = message.deepCopy();
+    boolean upgraded = false;
+
+    if (forceUpgrade || minor < 0) {
+      minor = 0;
+      patch = 0;
+      upgraded = true;
+    }
+
     if (minor < 3) {
```

Comment:
if (major == 1 && minor < 3)

Reply:
Major is always 1 (assertion at top of code block)